sklearn/metrics/_classification.py

```diff
@@ -506,6 +506,104 @@ def multilabel_confusion_matrix(y_true, y_pred, sample_weight=None,
     return np.array([tn, fp, fn, tp]).T.reshape(-1, 2, 2)
 
 
+def most_confused_classes(y_true, y_pred, labels=None, max_rows=None):
+    """Most confused classes.
+
+    In multiclass and binary classification, this function computes the classes
+    which have been confused the most often by the classifier.
+
+    Parameters
+    ----------
+
+    y_true: array, shape = [n_samples]
+        Ground truth (correct) target values.
+
+    y_pred : array, shape = [n_samples]
+        Estimated targets as returned by a classifier.
+
+    labels : array, shape = [n_classes], optional
+        List of labels to index the confusion matrix. This may be used to reorder
+        or select a subset of labels.
+        If none is given, those that appear at least once
+        in ``y_true`` or ``y_pred`` are used in sorted order.
+
+    max_rows : int, optional
+        Maximum number of pairs of classes to return. Corresponds to the
+        number of rows in most_confused.
+        If none is given, all rows are returned.
+
+    Returns
+    -------
+    most_confused : ndarray of shape (n_pairs, 3)
+        The most confused classes list, with each row containing:
+        [class_1, class_2, count]
+        where count denotes the number of times class_1 was misclassified as class_2
+        and n_pairs is the number of pairs of classes confused by the classifier.
+        Or, if there is no misclassification, returns an empty ndarray.
+
+    See also
+    --------
+    confusion_matrix
+
+    Examples
+    --------
+    >>> from sklearn.metrics import most_confused_classes
+    >>> y_true = [2, 0, 1, 2, 1, 2]
+    >>> y_pred = [0, 0, 2, 0, 1, 1]
+    >>> most_confused_classes(y_true, y_pred)
+    array([[2, 0, 2],
+           [2, 1, 1],
+           [1, 2, 1]], dtype=int64)
+
+    With labels:
+    >>> y_true = ["cat", "ant", "cat", "cat", "ant", "bird"]
+    >>> y_pred = ["ant", "ant", "cat", "cat", "ant", "cat"]
+    >>> most_confused_classes(y_true, y_pred, labels=["bird", "ant", "cat"])
+    array([[2, 1, 1],
+           [0, 2, 1]], dtype=int64)
+
+    Keep only two rows:
+    >>> from sklearn.metrics import most_confused_classes
+    >>> y_true = [2, 0, 1, 2, 1, 2]
+    >>> y_pred = [0, 0, 2, 0, 1, 1]
+    >>> most_confused_classes(y_true, y_pred, max_rows=2)
+    array([[2, 0, 2],
+           [2, 1, 1]], dtype=int64)
+    """
+    # If there is no mistake, we return an empty array
+    if np.array_equal(y_true, y_pred):
+        return np.empty(0)
+
+    if max_rows is not None and max_rows < 1:
+        raise ValueError('max_rows must be an int >= 1')
+
+    # confusion_matrix checks for input errors
+    cm = confusion_matrix(y_true, y_pred, labels=labels)
+
+    # Initialize the most_confused_dict dictionary
+    most_confused_dict = {}
```

Comment:
Try using numpy operations like np.argsort or np.argpartition over cm.


I'd start by clearing the diagonal and then finding the indices of the highest values (in the flattened array, which numpy then provides functions to convert back to the 2d indices)